opensearch_single_kernel/managers/lock.py

```diff
@@ -0,0 +1,507 @@
+#!/usr/bin/env python3
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""OpenSearch Lock manager.
+
+Ensure that only one node (re)starts, joins the cluster, or leaves the cluster at a time.
+
+The workflow logic goes alongside the following:
+
+1. If there are opensearch online nodes:
+   a) the node requesting the lock attempts to create a doc with the name of the node
+   b) if it succeeds => the unit gets the lock
+   c) if it fails => the unit doesn't get the lock as it is held by another unit
+   d) when the unit completes with their locked operation => releases the lock => deletes the doc
+2. if there are no online nodes:
+   a) we make use of a flag in the relation data
+   b) we check on the existence of the flag to know if the lock is held or not
+   c) if not there => we set the lock (flag in the peer rel data)
+   d) we release the lock by removing that flag from the rel data
+"""
+import json
+import os
+from typing import List, Optional
+
+import ops
+
+from opensearch_single_kernel.common.client import OpenSearchClient
+from opensearch_single_kernel.common.constants import DeploymentType, Scope, StartMode
+from opensearch_single_kernel.common.exceptions import OpenSearchHttpError
+from opensearch_single_kernel.core.models import DeploymentDescription
+from opensearch_single_kernel.core.state import ClusterState
+from opensearch_single_kernel.managers.base import BaseManager
+from opensearch_single_kernel.utils.helpers import format_unit_name
+from opensearch_single_kernel.utils.topology import ClusterTopology
+from opensearch_single_kernel.workload.base import BaseWorkload
+
+
+class PeerLockManager(BaseManager):
+    """Fallback lock when all units of OpenSearch are offline."""
+
+    def __init__(self, state: ClusterState, workload: BaseWorkload):
+        self.state = state
+        self.workload = workload
+
+    @property
+    def acquired(self) -> bool:
+        """Attempt to acquire lock.
+
+        Returns:
+            Whether lock was acquired
+        """
+        if not self._relation:
+            return False
+
+        self._relation.data[self.state.opensearch_unit.unit]["lock-requested"] = json.dumps(True)
+
+        if self.state.opensearch_unit.is_app_leader:
+            self.logger.debug("[Node lock] Requested peer lock as leader unit")
+            # A separate relation-changed event won't get fired
+            self._on_peer_relation_changed()
+
+        if self._unit_with_lock != self.state.opensearch_unit.unit_name:
+            self.logger.debug(
+                f"[Node lock] Not acquired. Unit with peer databag lock: {self._unit_with_lock}"
+            )
+            return False
+
+        if (
+            self.state.opensearch_unit.is_app_leader
+            and self._relation.data[self.state.opensearch_application.app][
```

Comment:
```suggestion
             self.state.server.is_app_leader 
         and self._relation.data[self.state.application.app][
```
self.state.opensearch_unit is not defined.